File: IEEE/Stones.py
# a simple parser for python. use get_number() and get_word() to read
import numpy
def parser():
    while 1:
        data = list(input().split(' '))
        for number in data:
            if len(number) > 0:
                yield(number)   

input_parser = parser()

def get_word():
    global input_parser
    return next(input_parser)

def get_number():
    data = get_word()
    try:
        return int(data)
    except ValueError:
        return float(data)

# numpy and scipy are available for use
import numpy
import scipy

def alice_wins_probability(R1, B1, R2, B2):
    # Initialize DP table
    dp = [[[[0.0 for _ in range(B2 + 1)] for _ in range(R2 + 1)] for _ in range(B1 + 1)] for _ in range(R1 + 1)]
    
    # Base case: If Bob is out of stones, Alice wins
    for aR in range(R1 + 1):
        for aB in range(B1 + 1):
            dp[aR][aB][0][0] = 1.0  # Alice wins if Bob has no stones left

    # Base case: If Alice is out of stones, she loses
    for bR in range(R2 + 1):
        for bB in range(B2 + 1):
            dp[0][0][bR][bB] = 0.0  # Alice loses if she has no stones left

    # Fill the DP table
    for aR in range(R1 + 1):
        for aB in range(B1 + 1):
            for bR in range(R2 + 1):
                for bB in range(B2 + 1):
                    if aR == 0 and aB == 0:
                        continue  # Skip, already handled
                    if bR == 0 and bB == 0:
                        continue  # Skip, already handled
                    
                    # Alice's turn
                    total = 0.0
                    if aR > 0:  # Alice can choose red
                        total += (dp[aR - 1][aB][bR][bB] * (bR / (bR + bB)))  # Bob guesses red and is correct
                        total += (dp[aR][aB][bR - 1][bB] * (1 - (bR / (bR + bB))))  # Bob guesses red and is wrong
                    
                    if aB > 0:  # Alice can choose blue
                        total += (dp[aR][aB - 1][bR][bB] * (bB / (bR + bB)))  # Bob guesses blue and is correct
                        total += (dp[aR][aB][bR][bB - 1] * (1 - (bB / (bR + bB))))  # Bob guesses blue and is wrong
                    
                    dp[aR][aB][bR][bB] = total

    # The answer we want is dp[R1][B1][R2][B2]
    return dp[R1][B1][R2][B2]

# Input reading
import sys
input = sys.stdin.read
R1, B1, R2, B2 = map(int, input().strip().split())

# Calculate probability
result = alice_wins_probability(R1, B1, R2, B2)

# Print result
print(f"{result:.9f}")

File: IEEE/Stones.cpp
#include<bits/stdc++.h>
using namespace std;

double dp[41][41][41][41]; // Memoization table

double calculate_probability(int R1, int B1, int R2, int B2) {
    if (dp[R1][B1][R2][B2] >= 0) {
        return dp[R1][B1][R2][B2];
    }

    if (R1 == 0 && B1 == 0) return dp[R1][B1][R2][B2] = 0.0;
    if (R2 == 0 && B2 == 0) return dp[R1][B1][R2][B2] = 1.0;

    double total_prob = 0.0;
    int total_outcomes = 0;

    if (R1 > 0) {
        if (R2 > 0) {
            total_prob += calculate_probability(R1 - 1, B1, R2 - 1, B2); // Alice loses a red
            total_outcomes++;
        }
        if (B2 > 0) {
            total_prob += calculate_probability(R1 - 1, B1, R2, B2 - 1); // Alice loses a red
            total_outcomes++;
        }
    }

    // Alice chooses blue
    if (B1 > 0) {
        // Bob guesses red
        if (R2 > 0) {
            total_prob += calculate_probability(R1, B1 - 1, R2 - 1, B2); // Alice loses a blue
            total_outcomes++;
        }
        // Bob guesses blue
        if (B2 > 0) {
            total_prob += calculate_probability(R1, B1 - 1, R2, B2 - 1); // Alice loses a blue
            total_outcomes++;
        }
    }

    // Average the probabilities
    if (total_outcomes > 0) {
        return dp[R1][B1][R2][B2] = total_prob / total_outcomes;
    }

    return dp[R1][B1][R2][B2] = 0.0; // No valid moves
}

int main() {
    int R1, B1, R2, B2;
    cin >> R1 >> B1 >> R2 >> B2;

    // Initialize the memoization table with -1
    for (int i = 0; i <= 40; i++) {
        for (int j = 0; j <= 40; j++) {
            for (int k = 0; k <= 40; k++) {
                for (int l = 0; l <= 40; l++) {
                    dp[i][j][k][l] = -1.0;
                }
            }
        }
    }

    double result = calculate_probability(R1, B1, R2, B2);
    cout << fixed << setprecision(6) << result << endl;

    return 0;
}
